lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgCommandStore.scala

```diff
@@ -107,6 +110,159 @@ class PostgresKgCommandStore @Inject()(configProvider: PostgresStoreConfigProvid
+        if (delta < convergenceThreshold) {
+          return
+        }
+      }
+    }
+
+    private def writeNodeLabelPageRank(convergenceThreshold: Double = 0.0000001, dampingFactor: Double = 0.85, maxIterations: Int = 20): Unit = {
+      val numNodeLabelsAction = nodeLabels.size.result
+      val initializeNodeLabelPageRanksAction = numNodeLabelsAction.flatMap {
+        case (numNodeLabels) => nodeLabels.map(_.pageRank).update(Some((1.0 / numNodeLabels).toFloat))
+      }
+      val numNodes = runSyncTransaction(DBIO.sequence(List(initializeNodeLabelPageRanksAction, writeNodeLabelDegreesAction)))(0)
+
+      if (numNodes == 0) {
+        return
+      }
+
+      for (_ <- 1 to maxIterations) {
+        val delta = runSyncTransaction(writeNodeLabelPageRankAction(dampingFactor.toFloat))(1).asInstanceOf[Number].doubleValue
+
+        if (delta < convergenceThreshold) {
+          return
+        }
+      }
+    }
+
+    private def writeNodePageRankAction(dampingFactor: Float) = {
+      val temporaryTableName = "temp_node_page_rank"
+      DBIO.sequence(List(
+        // Calculates new page rank for each node and saves in a temporary table
+        sqlu"""
+            SELECT n.id as node_id, ${dampingFactor} * sum(n_neighbor.page_rank / n.out_degree) + (1 - ${dampingFactor}) as page_rank
+            INTO #$temporaryTableName
+            FROM node n
+            JOIN edge e
+            ON e.object_node_id = n.id
+            JOIN node n_neighbor
+            ON n_neighbor.id = e.subject_node_id
+            GROUP BY n.id
+          """,
+        // Check if page rank has converged by calculating the difference
+        //  between the new page ranks and the current page ranks
+        sql"""
+           SELECT sqrt(sum((new.page_rank - cur.page_rank)^2))
+           FROM #$temporaryTableName new
+           JOIN node cur
+           ON cur.id = new.node_id
+         """.as[Double].head,
+        // Update the nodes with the new page ranks
+        sqlu"""
+            UPDATE node
+            SET
+              page_rank = new.page_rank
+            FROM #$temporaryTableName new
+            WHERE node.id = new.node_id
+          """,
+        // Drop the table
+        sqlu"DROP TABLE #$temporaryTableName"
+      ))
+    }
+
+    private def writeNodeLabelPageRankAction(dampingFactor: Float) = {
+      val temporaryTableName = "temp_node_label_page_rank"
+      DBIO.sequence(List(
+        sqlu"""
+            SELECT nl.label as label, ${dampingFactor} * sum(nl_neighbor.page_rank / nl.out_degree) + (1 - ${dampingFactor}) as page_rank
+            INTO #$temporaryTableName
+            FROM node_label nl
+            JOIN node_label_edge e
+            ON e.object_node_label_label = nl.label
+            JOIN node_label nl_neighbor
+            ON nl_neighbor.label = e.subject_node_label_label
+            GROUP BY nl.label
+          """,
+        sql"""
+           SELECT sqrt(sum((new.page_rank - cur.page_rank)^2))
```

Comment:
Add a RETURNING clause to UPDATE below, instead of doing select?

Reply:
Seems aggregate functions are not allowed in the RETURNING clause